apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/ActivationType.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.agent.configuration;
+
+import co.elastic.apm.agent.impl.ElasticApmTracer;
+import co.elastic.apm.agent.impl.GlobalTracer;
+import co.elastic.apm.agent.util.PrivilegedActionUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.lang.management.ManagementFactory;
+import java.util.List;
+
+public enum ActivationType {
+    K8S_ATTACH, // https://github.com/elastic/apm-mutating-webhook
+    PROGRAMMATIC_SELF_ATTACH, // ElasticApmAttacher.attach();
+    JAVAAGENT_FLAG, // -javaagent:... command-line option used
+    AWS_LAMBDA_LAYER, // Only if installed by using layers (as other metadata already identifies lambda)
+    ENV_ATTACH, // JAVA_TOOL_OPTIONS env var used to specify -javaagent option
+    APM_AGENT_ATTACH_CLI, // 'java -jar apm-agent-attach-cli.jar ...' used
+    FLEET, // Fleet using 'java -jar apm-agent-attach-cli.jar ...' to attach (directly or through webhook)
+
+    AZURE_FUNCTIONS, //Not defined yet how this is identified
+    UNKNOWN;
```

Comment:
If you move this to the `apm-agent-common` module, it would be visible to attacher, attacher-cli and core agent, which may improve reuse and save some conversions. You should be able to use it directly in `CoreConfiguration` and benefit from the automatic mapping.

Reply:
I deliberately kept the `ActivationType` - which is the type in the spec and reported by the agent - separate from `CoreConfiguration.ActivationMethod` because the reported spec/type has more options than the possible values passed to the agent, eg `ENV_ATTACH` is something that should be inferred by the agent by looking at the env rather than a passed value. I'm not wedded to this separation, but I still currently think it's a better option

I can move it to common, though I don't think anything else needs it so not sure it's better to widen the scope

Reply:
I am going to go with your recommendation, but I do still think these are logically two different things - an input type, and a reported type - which are connected rather than the same thing. But I agree it is simpler if I use one enum

Reply:
Had a go at this, but the class as is depends on core classes (like `ElasticApmTracer`) and trying to avoid the dependencies makes for a ridiculously convoluted implementation. And of course we can't have `common` depending on `core`

Reply:
Sorry, I was too haste to 👍 yesterday. I only meant adding a bare enum to the common module, without any dependency whatsoever. Any discovery logic can (and should) be part of all other metadata discovery.
Besides, looking at the `PrivilegedActionUtils` imports, it would perfectly fit the common module, though this is not required here if the discovery stays in core.